src/Preprocessor.zig

```diff
@@ -731,136 +655,278 @@ fn expandFunc(pp: *Preprocessor, source: *ExpandBuf, start_index: *usize, macro:
-    if (index == 0) return args;
-    // TODO this is a mess
-    var commas_seen: usize = 0;
-    var i: usize = 0;
+    // collect the arguments.
     var parens: u32 = 0;
-    while (i < args.len) : (i += 1) {
-        switch (args[i].id) {
-            .l_paren => parens += 1,
-            .r_paren => parens -= 1,
-            else => if (parens != 0) continue,
+    var args = MacroArguments.init(pp.comp.gpa);
+    errdefer args.deinit();
+    var curArgument = std.ArrayList(Token).init(pp.comp.gpa);
+    defer curArgument.deinit();
+    var done = false;
+    while (!done) {
+        var tok = try nextBufToken(tokenizer, buf, start_idx, end_idx, extend_buf);
+        switch (tok.id) {
+            .comma => {
+                if (parens == 0) {
+                    try args.append(curArgument.toOwnedSlice());
+                } else{
+                    try curArgument.append(tok);
+                }
+            },
+            .l_paren => {
+                try curArgument.append(tok);
+                parens += 1;
+            },
+            .r_paren => {
+                if (parens == 0) {
+                    try args.append(curArgument.toOwnedSlice());
+                    break;
+                } else {
+                    try curArgument.append(tok);
+                    parens -= 1;
+                }
+            },
+            .eof => {
+                tokenizer.index = initial_tokenizer_index;
+                end_idx.* = old_end;
+                try pp.comp.diag.add(.{ .tag = .unterminated_macro_arg_list, .loc = name_tok.loc });
+                return null;
+            },
+            else => {
+                try curArgument.append(tok);
+            }
         }
-        if (parens == 0 and args[i].id == .comma) commas_seen += 1;
-        if (commas_seen == index) return args[i + 1 ..];
     }
-    return args[i..];
+
+    return args;
 }
 
-/// get argument at index from a list of tokens.
-fn argSlice(args: []const Token, index: u32) []const Token {
-    // TODO this is a mess
-    var commas_seen: usize = 0;
-    var i: usize = 0;
-    var parens: u32 = 0;
-    while (i < args.len) : (i += 1) {
-        switch (args[i].id) {
-            .l_paren => parens += 1,
-            .r_paren => parens -= 1,
-            else => if (parens != 0) continue,
-        }
-        if (parens == 0 and args[i].id == .comma) {
-            if (index == 0) return args[0..i];
-            commas_seen += 1;
-            continue;
-        }
-        if (commas_seen == index) for (args[i..]) |a_2, j| {
-            if (parens == 0 and a_2.id == .comma) {
-                return args[i..][0..j];
+fn expandMacroExhaustive(pp: *Preprocessor, tokenizer: *Tokenizer, buf: *ExpandBuf, start_idx: usize, end_idx: usize, extend_buf: bool) Error!void {
+    var moving_end_idx = end_idx;
+    var advance_index: usize = 0;
+    // rescan loop
+    var do_rescan = true;
+    while (do_rescan) {
+        do_rescan = false;
```

Comment:
This logic looks like it could be simplified but I can't reason about it right now.

A slight improvement which allows simplifying other parts of the function would be to move the condition from the end into a continue expression:
```suggestion
    var do_rescan = true;
    while (do_rescan) : {
        // last phase before rescan: remove placeholder tokens
        // NOTE: only do this if there were expansion (i.e. do_rescan is true)
        if (do_rescan) {
            var i: usize = start_idx;
            while (i < buf.items.len) {
                const tok = &buf.items[i];
                switch (tok.id) {
                    .empty_arg => {
                        _ = buf.orderedRemove(i);
                        moving_end_idx -= 1;
                    },
                    else => i += 1,
                }
            }
        }
    } {
        do_rescan = false;
```
You'll probably want to double check these changes for correctness before adding them.